Add Individual Subscriber Stats

TimBroddin opened the pull request:
![CleanShot 2023-08-15 at 11 17 07@2x](https://github.com/Automattic/wp-calypso/assets/528287/f3497c26-1de0-4ec0-af84-524e70971d0b)


## Proposed Changes

This PR adds the boxes for individual subscriber stats to the subscriber detail page.

- It adds an `individual-subscriber-stats` flag, that's only enabled in development
- It adds a `<SubscriberStatsCard />` component, which is closely related to the HighlightCard component, but adds support for non-numerical values, loading state & tooltip support.
- It adds the `useSubscriberStatsQuery` react-query
- It adds the `<SubscriberStats />` component, which calls the query and renders the 3 boxes.
- It adds the needed icons

### Housekeeping
- This PR fixes how props are passed to the pages inside `subscribers/controller.js`. Previously it used `context.path`, which also includes the query string. By doing `const userId = path.split( '/' ).pop();` on `/subscribers/123456?page=1`, this would end up with a userId with value `123456?page=1`. I updated this to use `context.params`, which seems like the way to do this proper.
- This PR creates an export for { useSubscriberDetailsQuery } inside `queries/index.ts`, as this is how we've exported/imported queries before. It also updates the import inside `subscriber-details-page.tsx`
- This PR creates an export for { useSubscriptionPlans } inside `hooks/index.ts`, for the same reason. It also updates the imports.

## Testing Instructions

- First, apply this PR to your sandbox: D118596-code
- Sandbox `public-api.wordpress.com` 
- Apply this PR & `yarn start`
- Visit `Users > Subscribers`
- Click an external user, verify that the subscriber stats load
- Click on a WPCOM user, verify that the subscriber stats load

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

TimBroddin committed:
Add individual subscriber stats cards, currently behind a feature flag.

TimBroddin committed:
Add individual subscriber stats cards, currently behind a feature flag.

phcp commented on client/my-sites/subscribers/components/subscriber-stats-card/subscriber-stats-card.tsx:
```diff
@@ -0,0 +1,70 @@
+const SubscriberStatsCardValue = ( {
+	value,
+	isLoading,
+}: Pick< SubscriberStatsCardProps, 'value' | 'isLoading' > ) => {
```
Nice! TIL

phcp commented on client/my-sites/subscribers/components/subscriber-stats-card/subscriber-stats-card.tsx:
```diff
@@ -0,0 +1,70 @@
+		<Card className="highlight-card subscriber-stats-card">
+			<div className="highlight-card-icon subscriber-stats-card__icon">{ icon }</div>
+			<div className="highlight-card-heading subscriber-stats-card__heading">
+				{ heading }{ ' ' }
```
It would be cleaner to use margin instead of space.

phcp commented on client/my-sites/subscribers/components/subscriber-stats/icons/chart.tsx:
```diff
@@ -0,0 +1,11 @@
+import { SVG, Path } from '@wordpress/primitives';
+
+const Chart = (
```
This one is available in the [Icons library](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library) as `chartBar`, but not sure if worth it to use just this one from the library and the others from the SVG (at least I couldn't find the others on the library).

phcp commented on client/my-sites/subscribers/components/subscriber-stats/subscriber-stats.tsx:
```diff
@@ -0,0 +1,63 @@
+		userId
+	);
+
+	const openRate = useMemo( () => {
+		if ( subscriberStats && subscriberStats.emails_sent ) {
+			return ( ( subscriberStats.unique_opens / subscriberStats.emails_sent ) * 100 ).toFixed();
+		}
+		return 0;
+	}, [ subscriberStats ] );
+
+	const clickRate = useMemo( () => {
+		if ( subscriberStats && subscriberStats.emails_sent ) {
+			return ( ( subscriberStats.unique_clicks / subscriberStats.emails_sent ) * 100 ).toFixed();
+		}
+		return 0;
+	}, [ subscriberStats ] );
```
These functions share the same logic, it could be moved to a small helper function to calculate the rate.

phcp reviewed (approved):
Great job @TimBroddin! Code looks good and it's working as expected, just added some minor comments, but feel free to move forward with the PR.

TimBroddin committed:
Refactor subscriber stats open and click rates calculation

This commit removes repetitive rate calculation code in the subscriber stats component (open and click rates) and replaces it with a reusable hook `useSubscriberStatsRate`. It increases maintainability, since changes to the rate calculation now only need to be implemented in one place. One icon file `chart.tsx` was removed as it's replaced by using icon from WordPress icons directly. A dedicated type `SubscriberStats` was moved into a central types file to be re-used across components.

TimBroddin commented:
Thanks for the review @phcp, you caught some nice things that could use improvement. I've made all the suggested changes!

TimBroddin committed:
Enable individual subscriber stats in production

TimBroddin committed:
Add function to format subscriber rates

Added a function to handle formatting of subscriber stats rates in `useSubscriberStatsRate.ts`. The function ensures that integer rates are returned as integers and decimal rates are rounded to 1 decimal place. This change provides a consistent data type for the returned rates and improves readability.

phcp reviewed (approved):
Looks good!

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8720494

Thank you @TimBroddin for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8720494).
